### PR TITLE
Latent projection

### PR DESCRIPTION
**Pull request by fweber144**
This PR adds support for projpred's latent projection which has just been merged into projpred's `master` branch (see stan-dev/projpred#372).

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Thanks! When will the new projpred be on CRAN? I want to keep the brms master branch CRAN releasable at any time and would not want to merge this PR if it prevented releasing brms for now.

**Comment by fweber144** [comment]
> When will the new projpred be on CRAN?

I don't know yet, but the brms changes suggested here don't make brms dependent on the projpred version including the latent projection. Of course, that updated projpred version is needed for the latent projection, but all other projpred features are not affected. So even after merging this PR here, an older projpred version which excludes the latent projection can be used (for all features other than the latent projection). In particular, I was able to run brms's `R CMD check` with the current CRAN version of projpred. So concerning your comment

> I want to keep the brms master branch CRAN releasable at any time and would not want to merge this PR yet if it prevented releasing brms for now.

I think this PR should not prevent brms from being released on CRAN.

This is kind of a chicken-or-egg problem: Within projpred, we need to insert the brms version which supports the latent projection (currently, this is highlighted by a `TODO` comment in projpred, see https://github.com/stan-dev/projpred/pull/372#discussion_r1093152256). On the other hand, as you said, it would be desirable to have the latent projection in projpred's CRAN version before merging this brms PR here. So I thought the best solution would be to merge this brms PR and afterwards insert the current (GitHub) version number of brms into projpred. Even if brms would be published on CRAN before the projpred CRAN version includes the latent projection, users who might want to use the latent projection are instructed to use a projpred version > 2.3.0 (the current CRAN version) in the brms `NEWS` file (see https://github.com/paul-buerkner/brms/pull/1451/files#diff-51920e95310ebfbc1ae31709f3b95f89afffbf4f1a6e38e8b2b406e2fb6197ea), so they should realize that they need projpred's GitHub version (until the latent projection is included in a projpred CRAN version). Would you prefer to add such an instruction (to use a projpred version > 2.3.0) also to the documentation of argument `latent`?

Some technical details: The combination of this PR with projpred <= 2.3.0 is possible because `projpred::init_refmodel()`'s ellipsis (`...`, also exists in projpred <= 2.3.0) swallows up argument `latent` and any other arguments related to the latent projection that might be passed to `get_refmodel.brmsfit()`. So if users specify argument `latent` or other arguments related to it, these arguments simply become ineffective.

**Comment by paul-buerkner** [comment]
Makes sense. I will review this PR and then merge.

**Comment by fweber144** [comment]
Thanks! In any case, I'd like to have the latent projection in projpred's CRAN version soon, so it shouldn't take too long until it is available there.

File: R/projpred.R
Comment (paul-buerkner) [inline R/projpred.R:16]: In which cases would this have to be added manually?

File: R/projpred.R
Comment (paul-buerkner) [inline R/projpred.R:55]: I prefer seed being kind of last. Can we do `latent, dis, brms_seed, ...` order?

File: R/projpred.R
Comment (paul-buerkner) [inline R/projpred.R:206]: this looks like a lot of code overhead just for the cumulative family. Why is this necessary? 

And if it is necessary, can we perhaps move it to its own function to keep the code here a bit cleaner? I would like special cases to be stored away tidily rather than being around like this in the main function.

**Review by paul-buerkner — CHANGES_REQUESTED** [review]
I have some small comments to address before merging.

File: R/projpred.R
Comment (fweber144) [reply R/projpred.R:16]: For the latent projection with the probit or the logit link, projpred can use an internal default (constantly 1 and 1.6, respectively), but for other links, there is no such internal default. Perhaps users can come up with some reasonable `dis` value(s) in the future, so it's better to keep this flexible. Apart from that, `dis` is an argument of `init_refmodel()` and previously, it wasn't possible to specify it when using `get_refmodel.brmsfit()`, so already from a conceptual point of view, I think it makes sense to keep this flexible.

File: R/projpred.R
Comment (fweber144) [reply R/projpred.R:55]: Sure. Perhaps `dis, latent, brms_seed, ...` since `dis` is a long-standing argument of `init_refmodel()` and `latent` is a new one?

File: R/projpred.R
Comment (fweber144) [reply R/projpred.R:206]: It's necessary to define argument `latent_ilink` which is a new argument within projpred, see [here](https://github.com/stan-dev/projpred/blob/543a7c3a93fcff12b2cf5fe3982323592ae4c729/R/extend_family.R#L26-L32) and [here](https://github.com/stan-dev/projpred/blob/543a7c3a93fcff12b2cf5fe3982323592ae4c729/R/extend_family.R#L131-L154). For the cumulative family, `latent_ilink` is basically `inv_link_cumulative()`, but the draws can be clustered or thinned (that's why I used `S_agg` in the inline comments).

Comment (fweber144) [reply R/projpred.R:206]: > And if it is necessary, can we perhaps move it to its own function to keep the code here a bit cleaner?

Yes, I can try to define `latent_ilink_tmp` outside of `get_refmodel.brmsfit()`. That will require handing over several objects (which is the reason why I didn't define it outside of `get_refmodel.brmsfit()` until now). I'll check how many and which objects are concerned.

File: R/projpred.R
Comment (paul-buerkner) [reply R/projpred.R:16]: I see. Makes sense.

File: R/projpred.R
Comment (paul-buerkner) [reply R/projpred.R:55]: okay.

File: R/projpred.R
Comment (paul-buerkner) [reply R/projpred.R:206]: Thank you! Perhaps just passing the brms object (and some projpred stuff) would be sufficient?

Comment (fweber144) [reply R/projpred.R:206]: Done (commit https://github.com/paul-buerkner/brms/pull/1451/commits/0ebe5ef7c87171835603cc268876f3abdd9cf11f). Could you check whether my changes are ok?

> Perhaps just passing the brms object (and some projpred stuff) would be sufficient?

Unfortunately not. At least `resp` is required, too. And I also passed `family` and `bterms` to the new `get_latent_ilink_cumul()` function, for efficiency.

File: R/projpred.R
Comment (fweber144) [reply R/projpred.R]: Done (commit https://github.com/paul-buerkner/brms/pull/1451/commits/c6268cdc206fa3ef1be82a1582dcaeb453b3b749).

File: R/projpred.R
Comment (paul-buerkner) [reply R/projpred.R:206]: I made some minor changes mainly due to my taste of code appearence. Can you check that I didn't break anything. Once I got your okay, I will merge this PR.

Comment (fweber144) [reply R/projpred.R:206]: Looks good, shouldn't break anything. I just noticed that you also removed the `rm(draws_mat)` statement. I guess this can lead to unnecessary memory usage (because `draws_mat` will continue to live in the environment of the returned function), but if you say this extra memory usage is ok, I'm fine with merging.

**Comment by fweber144** [comment]
> ah I see. makes sense. I had overlooked this please add it back with a
comment for future me.

Done (commit fd090b8c).

**Review by paul-buerkner — APPROVED** [review]
looks good now. thank you!

**Comment by fweber144** [comment]
No problem. Thanks for merging!

**Pull request merged**
